lib/WP_Auth0_LoginManager.php

```diff
@@ -163,136 +170,124 @@ public function init_auth0() {
-      WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' => WP_Auth0_Api_Client::get_token()', $response );
-
-      error_log( $response->get_error_message() );
+    $exchange_resp_code = (int) wp_remote_retrieve_response_code( $exchange_resp );
+    $exchange_resp_body = wp_remote_retrieve_body( $exchange_resp );
 
-      throw new WP_Auth0_LoginFlowValidationException( $response->get_error_message() );
-    }
-
-    $data = json_decode( $response['body'] );
+    if ( 401 === $exchange_resp_code ) {
 
-    if ( isset( $data->access_token ) || isset( $data->id_token ) ) {
+      // Not authorized
+      WP_Auth0_ErrorManager::insert_auth0_error(
+        __METHOD__ . ' L:' . __LINE__,
+        __( 'An oauth/token call triggered a 401 response from Auth0. ', 'wp-auth0' ) .
+          __( 'Please check the Client Secret saved in the Auth0 plugin settings. ', 'wp-auth0' )
+      );
+      throw new WP_Auth0_LoginFlowValidationException( __( 'Not Authorized', 'wp-auth0' ), $exchange_resp_code );
 
-	    $decoded_token = JWT::decode(
-		    $data->id_token,
-		    $this->a0_options->get_client_secret_as_key(),
-		    array( $this->a0_options->get_client_signing_algorithm() )
-	    );
+    } else if ( empty( $exchange_resp_body ) ) {
 
-	    $data->id_token = null;
-	    $response = WP_Auth0_Api_Client::get_user(
-		    $this->a0_options->get( 'domain' ),
-		    WP_Auth0_Api_Client::get_client_token(),
-		    $decoded_token->sub
-	    );
+      // Unsuccessful for another reason
+      if ( $exchange_resp instanceof WP_Error ) {
+        WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' L:' . __LINE__, $exchange_resp );
+      }
 
-      if ( $response instanceof WP_Error ) {
-        WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' => WP_Auth0_Api_Client::get_user()', $response );
+      throw new WP_Auth0_LoginFlowValidationException( __( 'Unknown error', 'wp-auth0' ), $exchange_resp_code );
+    }
 
-        error_log( $response->get_error_message() );
+    $data = json_decode( $exchange_resp_body );
 
-        throw new WP_Auth0_LoginFlowValidationException( );
-      }
+    if ( empty( $data->access_token ) ) {
 
-      $userinfo = json_decode( $response['body'] );
-      if ( $this->login_user( $userinfo, $data->id_token, $data->access_token ) ) {
-        if ( ! empty( $state_decoded->interim ) ) {
-          include WPA0_PLUGIN_DIR . 'templates/login-interim.php';
-          exit();
-        } else {
-          if ( ! empty( $state_decoded->redirect_to ) && wp_login_url() !== $state_decoded->redirect_to ) {
-            $redirectURL = $state_decoded->redirect_to;
-          } else {
-            $redirectURL = $this->a0_options->get( 'default_login_redirection' );
-          }
+      // Look for clues as to what went wrong
+      throw new WP_Auth0_LoginFlowValidationException(
+        ! empty( $data->error_description ) ? $data->error_description : __( 'Unknown error', 'wp-auth0' ),
```

Comment:
Can you explain the line below this one please? I'm lost after the comma at the end.

Reply:
`WP_Auth0_LoginFlowValidationException` extends `Exception` which accepts a message as the first argument and a code as the second:

https://secure.php.net/manual/en/exception.construct.php

So this is trying to get a message to use on the first line and a code on the second ... tertiaries as arguments. I think the extemporaneous parentheses on that second arg are confusing so I'm removing them.

Reply:
I find the code easier to follow if you fetch the values before creating the exception, unless you have some performance impact for doing so.

```php
message = $data->error_description ) ? $data->error_description : __( 'Unknown error', 'wp-auth0' )
code = ! empty( $data->error ) ? $data->error : $exchange_resp_code
throw new WP_Auth0_LoginFlowValidationException(message, code)
```

Reply:
@lbalmaceda - Agree and fixed!